crm_salesperson_planner/models/crm_salesperson_planner_visit_template.py

```diff
@@ -0,0 +1,226 @@
+    alarm_ids = fields.Many2many(relation="visit_calendar_event_rel")
+    state = fields.Selection(
+        string="Status",
+        required=True,
+        copy=False,
+        tracking=True,
+        selection=[
+            ("draft", "Draft"),
+            ("in-progress", "In Progress"),
+            ("done", "Done"),
+            ("cancel", "Cancelled"),
+        ],
+        default="draft",
+    )
+    visit_ids = fields.One2many(
+        comodel_name="crm.salesperson.planner.visit",
+        inverse_name="visit_template_id",
+        string="Visit Template",
+    )
+    visit_ids_count = fields.Integer(
+        string="Number of Sales Person Visits", compute="_compute_visit_ids_count"
+    )
+    auto_validate = fields.Boolean(string="Auto Validate", default=True)
+    rrule_type = fields.Selection(default="daily", required=True,)
+    last_visit_date = fields.Date(
+        string="Last Visit Date", compute="_compute_last_visit_date", store=True
+    )
+    final_date = fields.Date(string="Repeat Until")
+    allday = fields.Boolean(default=True)
+    recurrency = fields.Boolean(default=True)
+
+    _sql_constraints = [
+        (
+            "crm_salesperson_planner_visit_template_name",
+            "UNIQUE (name)",
+            "The visit template number must be unique!",
+        ),
+    ]
+
+    def _compute_visit_ids_count(self):
+        visit_data = self.env["crm.salesperson.planner.visit"].read_group(
+            [("visit_template_id", "in", self.ids)],
+            ["visit_template_id"],
+            ["visit_template_id"],
+        )
+        mapped_data = {
+            m["visit_template_id"][0]: m["visit_template_id_count"] for m in visit_data
+        }
+        for sel in self:
+            sel.visit_ids_count = mapped_data.get(sel.id, 0)
+
+    @api.depends("visit_ids.date")
+    def _compute_last_visit_date(self):
+        for sel in self.filtered(lambda x: x.visit_ids):
+            sel.last_visit_date = sel.visit_ids.sorted(lambda x: x.date)[-1].date
+        self.filtered(lambda x: not x.visit_ids).write({"last_visit_date": False})
+
+    # overwrite
+    def _search(
+        self,
+        args,
+        offset=0,
+        limit=None,
+        order=None,
+        count=False,
+        access_rights_uid=None,
+    ):
+        return super(models.Model, self)._search(args)
+
+    @api.constrains("partner_ids")
+    def _constrains_partner_ids(self):
+        for sel in self:
+            if len(sel.partner_ids) > 1:
+                raise ValidationError(_("Only one customer is allowed"))
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            if vals.get("name", "/") == "/":
+                vals["name"] = self.env["ir.sequence"].next_by_code(
+                    "salesperson.planner.visit.template"
+                )
+        return super().create(vals_list)
+
+    # overwrite
+    def read(self, fields=None, load="_classic_read"):
```

Comment:
Why are you doing all of these overridings? I don't think they are needed, as these are not virtual records. Anyways, always that you make something controversial, document it.

Reply:
Better explained.